core/chains/evm/gas/universal_estimator.go

```diff
@@ -0,0 +1,413 @@
+	bigmath "github.com/smartcontractkit/chainlink-common/pkg/utils/big_math"
+
+	commonfee "github.com/smartcontractkit/chainlink/v2/common/fee"
+	feetypes "github.com/smartcontractkit/chainlink/v2/common/fee/types"
+	"github.com/smartcontractkit/chainlink/v2/core/chains/evm/assets"
+	evmclient "github.com/smartcontractkit/chainlink/v2/core/chains/evm/client"
+	"github.com/smartcontractkit/chainlink/v2/core/chains/evm/gas/rollups"
+	evmtypes "github.com/smartcontractkit/chainlink/v2/core/chains/evm/types"
+)
+
+// metrics are thread safe
+var (
+	promUniversalEstimatorGasPrice = promauto.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "gas_price_updater",
+		Help: "Sets latest gas price (in Wei)",
+	},
+		[]string{"evmChainID"},
+	)
+	promUniversalEstimatorBaseFee = promauto.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "base_fee_updater",
+		Help: "Sets latest BaseFee (in Wei)",
+	},
+		[]string{"evmChainID"},
+	)
+	promUniversalEstimatorMaxPriorityFeePerGas = promauto.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "max_priority_fee_per_gas_updater",
+		Help: "Sets latest MaxPriorityFeePerGas (in Wei)",
+	},
+		[]string{"evmChainID"},
+	)
+	promUniversalEstimatorMaxFeePerGas = promauto.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "max_fee_per_gas_updater",
+		Help: "Sets latest MaxFeePerGas (in Wei)",
+	},
+		[]string{"evmChainID"},
+	)
+)
+
+const (
+	MinimumBumpPercentage   = 10 // based on geth's spec
+	ConnectivityPercentile  = 85
+	BaseFeeBufferPercentage = 40
+)
+
+type UniversalEstimatorConfig struct {
+	BumpPercent  uint16
+	CacheTimeout time.Duration
+	EIP1559      bool
+
+	BlockHistorySize uint64
+	RewardPercentile float64
+	HasMempool       bool
+}
+
+type universalEstimatorClient interface {
+	SuggestGasPrice(ctx context.Context) (*big.Int, error)
+	FeeHistory(ctx context.Context, blockCount uint64, rewardPercentiles []float64) (feeHistory *ethereum.FeeHistory, err error)
+}
+
+type UniversalEstimator struct {
+	services.StateMachine
+
+	client  universalEstimatorClient
+	logger  logger.Logger
+	config  UniversalEstimatorConfig
+	chainID *big.Int
+
+	gasPriceMu sync.RWMutex
+	gasPrice   *assets.Wei
+
+	dynamicPriceMu sync.RWMutex
+	dynamicPrice   DynamicFee
+
+	priorityFeeThresholdMu sync.RWMutex
+	priorityFeeThreshold   *assets.Wei
+
+	l1Oracle rollups.L1Oracle
+
+	wg     *sync.WaitGroup
+	stopCh services.StopChan
+}
+
+func NewUniversalEstimator(lggr logger.Logger, client universalEstimatorClient, cfg UniversalEstimatorConfig, chainID *big.Int, l1Oracle rollups.L1Oracle) *UniversalEstimator {
+	return &UniversalEstimator{
+		client:   client,
+		logger:   logger.Named(lggr, "UniversalEstimator"),
+		config:   cfg,
+		chainID:  chainID,
+		l1Oracle: l1Oracle,
+		wg:       new(sync.WaitGroup),
+		stopCh:   make(chan struct{}),
+	}
+}
+
+func (u *UniversalEstimator) Start(context.Context) error {
+	return u.StartOnce("UniversalEstimator", func() error {
+		if u.config.BumpPercent < MinimumBumpPercentage {
```

Comment:
why do the config checks in the start function vs in the constructor when we pass them in initially?

Reply:
This is somewhat Go's idiomatic way of doing this. If you check across our code base we hardly ever return an error when calling a constructor. Instead, we batch-start services and collect the errors.
On a separate note, Amit made a point of moving this check in our config validations, but when I was writing the estimator my configs were abstracted completely so I wasn't able to do that. By the time I was done, it also seemed cleaner to me to keep the checks sandboxed on the component that needs them rather than having a centralized place where things don't make much sense. I don't have a strong preference, but ideally, I would like us to try and move away from the multi-declaration config mess we have now.